File: sample-chat-obj-c/sample-chat/Components/NewDialogScreen/EnterChatNameVC.m
//
//  EnterChatNameVC.m
//  sample-chat
//
//  Created by Injoit on 04.02.2020.
//  Copyright © 2020 Quickblox. All rights reserved.
//

#import "EnterChatNameVC.h"
#import "InputContainer.h"
#import "TitleView.h"
#import "ChatManager.h"
#import "DialogsViewController.h"
#import "ChatViewController.h"
#import "Log.h"
#import "UIView+Chat.h"
#import "UITextField+Chat.h"
#import "UIViewController+Alert.h"

NSString *const NAME_HINT = @"Must be in a range from 3 to 20 characters.";
NSString *const NAME_REGEX = @"^[^_]{3,19}$";
NSString *const NAME_TITLE = @"Chat Name";

@interface EnterChatNameVC () <InputContainerDelegate>
//MARK: - Properties
@property (strong, nonatomic) InputContainer *chatNameInputContainer;
@property (nonatomic, strong) TitleView *titleView;
@property (nonatomic, strong) ChatManager *chatManager;

@end

@implementation EnterChatNameVC
#pragma mark - Life Cycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.chatNameInputContainer = [[NSBundle mainBundle] loadNibNamed:@"InputContainer" owner:nil options:nil].firstObject;
    [self.chatNameInputContainer setupWithTitle:NAME_TITLE hint:NAME_HINT regexes:@[NAME_REGEX]];
    [self.view addSubview:self.chatNameInputContainer];
    self.chatNameInputContainer.translatesAutoresizingMaskIntoConstraints = NO;
    [self.chatNameInputContainer.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = YES;
    [self.chatNameInputContainer.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = YES;
    [self.chatNameInputContainer.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:16.0f].active = YES;
    [self.chatNameInputContainer inputTextfieldBecomeFirstResponder];
    [self.chatNameInputContainer layoutIfNeeded];
    self.chatNameInputContainer.delegate = self;
    
    self.titleView = [[TitleView alloc] init];
    self.navigationItem.titleView = self.titleView;
    [self setupNavigationTitle];
    
    self.chatManager = [ChatManager instance];
    
    UIBarButtonItem *createButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Finish"
                                                                         style:UIBarButtonItemStylePlain
                                                                        target:self
                                                                        action:@selector(createChatButtonPressed:)];
    self.navigationItem.rightBarButtonItem = createButtonItem;
    createButtonItem.tintColor = UIColor.whiteColor;
    self.navigationItem.rightBarButtonItem.enabled = NO;
    
    UIBarButtonItem *backButtonItem = [[UIBarButtonItem alloc] initWithImage:[UIImage imageNamed:@"chevron"]
                                                                       style:UIBarButtonItemStylePlain
                                                                      target:self
                                                                      action:@selector(didTapBack:)];
    
    self.navigationItem.leftBarButtonItem = backButtonItem;
    backButtonItem.tintColor = UIColor.whiteColor;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];

    if (QBChat.instance.isConnected == NO) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
}

#pragma mark - Setup

- (void)setupNavigationTitle {
    NSString *title = @"New Chat";
    NSString *numberUsers = [NSString stringWithFormat:@"%@ users selected", @(self.selectedUsers.count)];
    [self.titleView setupTitleViewWithTitle:title subTitle:numberUsers];
}

#pragma mark - Actions
- (IBAction)createChatButtonPressed:(UIButton *)sender {
    if (QBChat.instance.isConnected == NO) {
        [self showNoInternetAlertWithHandler:nil];
        return;
    }
    [self.chatManager.storage updateUsers:self.selectedUsers];
    NSString *chatName = self.chatNameInputContainer.text;
    sender.enabled = NO;
    __weak __typeof(self)weakSelf = self;
    [self.chatManager createGroupDialogWithName:chatName occupants:self.selectedUsers completion:^(NSError * _Nullable error, QBChatDialog * _Nullable createdDialog) {
        __typeof(weakSelf)strongSelf = weakSelf;
        if (!createdDialog || error) {
            sender.enabled = YES;
            [weakSelf showAlertWithTitle:nil
                                 message:error.localizedDescription
                                 handler:nil];
            return;
        }
        NSArray *controllers = strongSelf.navigationController.viewControllers;
        NSMutableArray *newStack = [NSMutableArray array];
        
        //change stack by replacing view controllers after DialogsVC with ChatVC
        for (UIViewController *controller in controllers) {
            [newStack addObject:controller];
            if ([controller isKindOfClass:[DialogsViewController class]]) {
                UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Chat" bundle:nil];
                ChatViewController *chatController = [storyboard instantiateViewControllerWithIdentifier:@"ChatViewController"];
                chatController.dialogID = createdDialog.ID;
                [newStack addObject:chatController];
                [strongSelf.navigationController setViewControllers:[newStack copy]];
                return;
            }
        }
    }];
}

- (void)didTapBack:(UIButton *)sender {
    [self.navigationController popViewControllerAnimated:YES];
}

#pragma mark - InputContainerDelegate
- (void)inputContainer:(nonnull InputContainer *)inputContainer didChangeValidState:(BOOL)isValid {
    self.navigationItem.rightBarButtonItem.enabled = isValid;
}

@end

File: sample-chat-swift/sample-chat-swift/Components/NewDialogScreen/EnterChatNameVC.swift
//
//  EnterChatNameVC.swift
//  sample-chat-swift
//
//  Created by Injoit on 10/9/19.
//  Copyright © 2019 Quickblox. All rights reserved.
//

import UIKit

class EnterChatNameVC: UIViewController {
    
    //MARK: - Properties
    lazy var chatNameInputContainer: InputContainer = {
        let chatNameInputContainer = InputContainer.loadNib()
        chatNameInputContainer.setup(title: .chatName,
                                     hint: .chatName,
                                     regexes: [.chatName])
        chatNameInputContainer.delegate = self
        return chatNameInputContainer
    }()
    var selectedUsers: [QBUUser] = []
    private let chatManager = ChatManager.instance
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.addSubview(chatNameInputContainer)
        chatNameInputContainer.translatesAutoresizingMaskIntoConstraints = false
        chatNameInputContainer.leftAnchor.constraint(equalTo: view.leftAnchor).isActive = true
        chatNameInputContainer.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16.0).isActive = true
        chatNameInputContainer.rightAnchor.constraint(equalTo: view.rightAnchor).isActive = true
        chatNameInputContainer.inputTextfield.becomeFirstResponder()
        chatNameInputContainer.layoutIfNeeded()
        
        let backButtonItem = UIBarButtonItem(image: UIImage(named: "chevron"),
                                             style: .plain,
                                             target: self,
                                             action: #selector(didTapBack(_:)))
        navigationItem.leftBarButtonItem = backButtonItem
        backButtonItem.tintColor = .white
        
        let createButtonItem = UIBarButtonItem(title: "Finish",
                                               style: .plain,
                                               target: self,
                                               action: #selector(createChatButtonPressed(_:)))
        navigationItem.rightBarButtonItem = createButtonItem
        createButtonItem.tintColor = .white
        navigationItem.rightBarButtonItem?.isEnabled = false
        
        title = "New Chat"
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        if QBChat.instance.isConnected == false {
            showNoInternetAlert(handler: nil)
            return
        }
    }
    
    //MARK: - Actions
    @objc func didTapBack(_ sender: UIBarButtonItem) {
        navigationController?.popViewController(animated: true)
    }
    
    @objc func createChatButtonPressed(_ sender: UIBarButtonItem) {
        if QBChat.instance.isConnected == false {
            showNoInternetAlert(handler: nil)
            return
        }
        let chatName = chatNameInputContainer.inputTextfield.text ?? "New Group Chat"
        sender.isEnabled = false
        chatManager.storage.update(users: selectedUsers)
        
        chatManager.createGroupDialog(withName: chatName,
                                      photo: nil,
                                      occupants: selectedUsers) { [weak self] (error, dialog) -> Void in
            guard let self = self else {return}
            guard let dialog = dialog,
                  let dialogID = dialog.id else {
                      if let error = error {
                          self.showAlertView(nil, message: error, handler: nil)
                      }
                      sender.isEnabled = true
                      return
                  }
            guard let navigationController = self.navigationController else {
                return
            }
            let controllers = navigationController.viewControllers
            var newStack = [UIViewController]()
            
            //change stack by replacing view controllers after ChatVC with ChatVC
            controllers.forEach {
                newStack.append($0)
                if $0 is DialogsViewController {
                    guard let chatController = Screen.chatViewController() else {
                        return
                    }
                    chatController.dialogID = dialogID
                    newStack.append(chatController)
                    navigationController.setViewControllers(newStack, animated: true)
                    return
                }
            }
        }
    }
}

extension EnterChatNameVC: InputContainerDelegate {
    func inputContainer(_ container: InputContainer, didChangeValidState isValid: Bool) {
        if isValid == false {
            navigationItem.rightBarButtonItem?.isEnabled = false
            return
        }
        navigationItem.rightBarButtonItem?.isEnabled = true
    }
}
